### ISSUE TITLE
[RFC] Repository and Inspector Factory

### ISSUE THREAD

**Issue by dantleech** [sulu-io/sulu-document-manager#11]
Currently we have a _single_ inspector class (inspector allows you to infer details about a document, e.g. locale, webspace, children, etc.) and there is no concept of reposiotries.

It would be a good idea, I think, to have,m potentially distinct Inspector classes for each entitiy.

Repositories are necessary in any ORM/ODM.

Both of these requirements can be met by creating decoupled services. **BUT** just imagine writing a command which much deal with 10 different types of entity, each of which needs its own inspector. You would need to inject 10 inspectors and 10 repositories. Not nice!.

I propose to create factories for inspectors and repositories.

``` php
$webspace = $inspectorFactory->getInspector($document)->getWebspace();
$repository = $repositoryFactory->getRepository(Class::class);
```

If no repository/inspector has been mapped, then generic instances will be returned.

and we could also extend the DocumentManager API, a la Doctrine:

``` php
$documentManager->getInspector($document)->getWebspace();
$fooDocuments = $documentManager->getRepository(Document::class)->findByFoo();
```

**Comment by danrot** [comment]
I just had a look at the Inspector, and I am asking myself why do we need it at all? Why are calls like https://github.com/sulu-io/sulu-document-manager/blob/f66314d4ba2441f96fef8103f657eb84bbd9637c/lib/DocumentInspector.php#L191 not handled by the document itself? That's also about encapsulation, I guess we don't want others to work directly with the node.

**Comment by dantleech** [comment]
Yeah, we can't gurantee that people implement, for example, the `UuidBehavior`, but internally we need this information. The document is basically unaware of the system, it does not _contain_ anything. It is assosiated with the node externally in tne the DocumentRegistry.

**Comment by dantleech** [comment]
Have been thinking about this and I think it makes sense.

`$dm->getInspector($document / $documentFqn)`
`$dm->getRepository($document / $documentFqn)`

I think the inspector is a good idea atm -- the alternative is to always load things during hydration which may not be required during the request and I think we should keep hydration as light-weight as possible.

One example which I encountered recently was getting the `resource-locator` for a route, other examples include getting the webspace name from the ContentDocument.

**Comment by webmozart** [comment]
Can you explain in more detail what you use inspectors for?

**Comment by dantleech** [comment]
Examples:
- $inspector->getWebspaceName($doc);
- $inspector->getResourceLocator($doc); // a sulu concept, one part of the URL for a route ../{locale}/{resource_locator}
- $inspector->getLocale($doc);
- $inspector->isManaged($doc);
- $inspector->getChildren($doc);
- $inspector->getParent($doc);

These values could be hydrated, but we can't force this upon a user and
it would be inefficient to hydate these "system" properties when they
might not be required.

In the Doctrine world (or at least the PHPCR-ODM world) such methods
would be part of the DocumentManager API, but this couples this mixes
concerns and doesn't allow for domain specialisation (the idea in this
issue).
